### PR TITLE
JavaScriptInvoke with JavaScriptRefNum types.

### PR DESCRIPTION
**Pull request by ggumen**
Adding support for JavaScriptRefNum types for JavaScriptInvoke (#450)
- Adding eggshell reader/writer for JavaScriptRefNum type.
- JavaScriptRefNum is shared among the same js object.
- Updating JavaScriptInvoke to support JavaScriptRefNum type.
- Adding tests for JavaScriptInvoke wiht JavaScriptRefNum and ensuring references can be shared.

### CODE REVIEW

File: source/io/module_javaScriptInvoke.js [hunk 1]
```diff
@@ -103,6 +102,61 @@
             return typeof value === 'number';
         };
 
+        var jsRefNumToJsObjectMap = new Map();
+        var jsObjectToJsRefNumMap = new Map();
+        var jsRefNumCookieCounter = 0;
+
+        var cacheRefNum = function (cookie, jsObject) {
+            jsRefNumToJsObjectMap.set(cookie, jsObject);
+            jsObjectToJsRefNumMap.set(jsObject, cookie);
```
Comment (cglzaguilar) [inline source/io/module_javaScriptInvoke.js]: Do we need to consider caching only objects and not scalars?

File: test-it/karma/fixtures/javascriptinvoke/ObjectType.via [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        JavaScriptInvoke(occurrence false error3 'NI_UseObjectFunction' length1 ref1)
+        JavaScriptInvoke(occurrence false error4 'NI_UseObjectFunction' length2 ref2)
+        
+        // Get the same js ref
+        JavaScriptInvoke(occurrence false error5 'NI_GetObjectFunction' ref3 name1) 
+        IsEQ(ref1 ref3 isSharedRef)
+    )
+) ) )
+
+enqueue(MyVI)
```
Comment (cglzaguilar) [inline test-it/karma/fixtures/javascriptinvoke/ObjectType.via]: Add a new line to the end of the file.

File: source/io/module_javaScriptInvoke.js [hunk 3]
```diff
@@ -102,29 +102,35 @@
...
             return refNumExists;
         };
 
-        var getCachedRefNum = function (jsObject) {
-            return jsObjectToJsRefNumMap.get(jsObject);
+        var getCachedRefNum = function (jsValue) {
+            return jsValueToJsRefNumCache.get(jsValue);
         };
 
         var getCachedJsObject = function (cookie) {
```
Comment (rajsite) [inline source/io/module_javaScriptInvoke.js]: Maybe getStoredRefNum is a more accurate name?

File: source/io/module_javaScriptInvoke.js (hunk 1 above)
Comment (ggumen) [reply source/io/module_javaScriptInvoke.js]: Yes! We discussed this in the past, but it seems I forgot to consider that case. Updated.

File: test-it/karma/fixtures/javascriptinvoke/ObjectType.via (hunk 2 above)
Comment (ggumen) [reply test-it/karma/fixtures/javascriptinvoke/ObjectType.via]: done.

File: test-it/karma/fixtures/javascriptinvoke/ObjectType.via [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        e(dv(.Boolean false) isNotEqual)
+        e(dv(.Boolean false) isNotANumPathRefnum1)
+        e(dv(.Boolean false) isNotANumPathRefnum2)
+        e(dv(.Boolean false) isSharedRef)
+        e(dv(.Boolean false) isSharedPrimRef)
+        e(.Occurrence occurrence)
+    )
+    clump (
+        // Get js object ref
+        JavaScriptInvoke(occurrence false error1 'NI_GetObjectFunction' ref1 name1)
```
Comment (sanmut) [inline test-it/karma/fixtures/javascriptinvoke/ObjectType.via]: Talked to Gandi about a couple of more test cases including checking for "not-a-refnum" on a ref before calling JSInvoke on that.

File: source/core/JavaScriptRef.cpp [hunk 5]
```diff
@@ -18,12 +18,21 @@
...
+    extern void jsIsNotAJavaScriptRefnum(Boolean*, JavaScriptRefNum*);
+}
+#endif
 
 VIREO_FUNCTION_SIGNATURE2(IsNotAJavaScriptRefnum, JavaScriptRefNum, Boolean)
 {
     JavaScriptRefNum* refnumPtr = _ParamPointer(0);
-    bool result = *refnumPtr == 0;  // this will be changed to call into JS to validate reference.
-    _Param(1) = result;
+    Boolean result = false;
```
Comment (cglzaguilar) [inline source/core/JavaScriptRef.cpp]: We can simplify this code as follows:
Boolean isNotARefnum = false;
jsIsNotAJavaScriptRefnum(&isNotARefnum, refnumPtr);
_Param(1) = isNotARefnum;

**Pull request merged**